marketdata/stream/client.go

```diff
@@ -65,63 +95,85 @@ type client struct {
 	in             chan []byte
 	subChanges     chan []byte
 
-	handlerMutex          sync.RWMutex
-	trades                []string
-	tradeHandler          func(trade Trade)
-	quotes                []string
-	quoteHandler          func(quote Quote)
-	bars                  []string
-	barHandler            func(bar Bar)
-	dailyBars             []string
-	dailyBarHandler       func(bar Bar)
+	trades    []string
+	quotes    []string
+	bars      []string
+	dailyBars []string
+
+	handler msgHandler
+
 	pendingSubChangeMutex sync.Mutex
 	pendingSubChange      *subChangeRequest
-}
 
-var _ StreamV2Client = (*client)(nil)
+	connCreator func(ctx context.Context, u url.URL) (conn, error)
+}
 
-// NewClient returns a new StreamV2Client that will connect to feed data feed
-// and whose default configurations are modified by opts.
-func NewClient(feed string, opts ...Option) StreamV2Client {
-	c := client{
-		feed:           feed,
+func newClient() *client {
+	return &client{
 		terminatedChan: make(chan error, 1),
 		subChanges:     make(chan []byte, 1),
 	}
-	o := defaultOptions()
-	o.apply(opts...)
-	c.configure(*o)
-	return &c
 }
 
 func (c *client) configure(o options) {
 	c.logger = o.logger
-	c.host = o.host
+	c.baseURL = o.baseURL
 	c.key = o.key
 	c.secret = o.secret
 	c.reconnectLimit = o.reconnectLimit
 	c.reconnectDelay = o.reconnectDelay
 	c.processorCount = o.processorCount
 	c.bufferSize = o.bufferSize
-	c.handlerMutex.Lock()
-	defer c.handlerMutex.Unlock()
 	c.trades = o.trades
-	c.tradeHandler = o.tradeHandler
 	c.quotes = o.quotes
-	c.quoteHandler = o.quoteHandler
 	c.bars = o.bars
-	c.barHandler = o.barHandler
 	c.dailyBars = o.dailyBars
-	c.dailyBarHandler = o.dailyBarHandler
+	c.connCreator = o.connCreator
+}
+
+type stocksClient struct {
+	*client
+
+	feed    string
+	handler *stocksMsgHandler
```

Comment:
I wish `handler` weren't handler because of the clash it causes with the embedded struct. Same for `cryptoClient`. I can't recommend a better name right now but perhaps you have a good idea.

Reply:
This was actually intentional. `sc.client.handler` is an interface pointing to `sc.handler`. `sc.client.handler` should never be used from a caller's perspective.